Security_Engine/Compute/CameraFieldOfView.cs

```diff
@@ -0,0 +1,351 @@
+            return visibleLine;
+        }
+
+        /***************************************************/
+
+        [Description("Create dictionary from line and obstacle.")]
+        [Input("line", "Line to create dictionary for.")]
+        [Input("obstacles", "Polyline objects that represents obstacles for lines.")]
+        [Output("dictionary", "Dictionary for lines and obstacles.")]
+        private static Dictionary<Line, Polyline> LineObstacleDictionary(this Line line, List<Polyline> obstacles)
+        {
+            Dictionary<Line, Polyline> lineObstDict = new Dictionary<Line, Polyline>();
+            if (obstacles.Count == 0)
+            {
+                lineObstDict.Add(line, null);
+            }
+            for (int i = 0; i < obstacles.Count; i++)
+            {
+                if (obstacles[i].IsContaining(new List<Point> { line.End }))
+                {
+                    lineObstDict.Add(line, obstacles[i]);
+                    return lineObstDict;
+                }
+                else if (i == obstacles.Count - 1)
+                {
+                    lineObstDict.Add(line, null);
+                }
+            }
+
+            return lineObstDict;
+        }
+
+        /***************************************************/
+
+        [Description("Create points chain for the camera field of view.")]
+        [Input("lineObstacledictionary", "Dictionary of visible lines and obstacles.")]
+        [Input("cameraLocation", "Location of the camera device.")]
+        [Input("radius", "Radius of the camera view cone.")]
+        [Output("pointsChain", "Points chain of the camera field of view.")]
+        private static List<Point> PointsChain(this List<Dictionary<Line, Polyline>> lineObstacledictionary, Point cameraLocation, double radius, double tolerance)
+        {
+            List<Point> pointsChain = new List<Point>();
+            pointsChain.Add(cameraLocation);
+            Point lastPoint = cameraLocation;
+            for (int i = 0; i < lineObstacledictionary.Count; i++)
+            {
+                Line line = lineObstacledictionary[i].Keys.First();
+                if (line.Start.IsEqual(cameraLocation))
+                {
+                    Point point = line.End;
+                    pointsChain.Add(point);
+                    lastPoint = point;
+                }
+                else
+                {
+                    Point pt1 = line.Start;
+                    Point pt2 = line.End;
+
+                    if (lastPoint.Distance(pt1) < lastPoint.Distance(pt2) && !((Math.Abs(lastPoint.Distance(cameraLocation) - radius) < tolerance) && (Math.Abs(pt2.Distance(cameraLocation) - radius) < tolerance)) && i > 0)
+                    {
+                        Polyline obst = lineObstacledictionary[i].Values.First();
+                        Polyline lastObst = lineObstacledictionary[i - 1].Values.First();
+                        if (!lastPoint.IsEqual(cameraLocation) && lastObst == obst)
```

Comment:
Again, tolerance should be passed across the entire method chain, reaching this point